backend/src/load.rs

```diff
@@ -0,0 +1,404 @@
+            let date = try!(parse_from_header(&date).or_else(|_| parse_from_filename(&filename)));
+
+            let test_name = &filename[0..filename.find("--").unwrap()];
+
+            let times = &contents["times"];
+            if test_name == "rustc" {
+                c_bootstraps += 1;
+                data_rustc.push(Data::new(date, &header, times, true));
+
+                for timing in times.members() {
+                    let crate_name = timing["crate"].as_str().unwrap().to_string();
+                    crate_list.insert(crate_name);
+                }
+            } else {
+                benchmarks.insert(test_name.to_string());
+                let index = data_benchmarks.iter().position(|benchmark: &Data| benchmark.date == date);
+                if let Some(index) = index {
+                    c_benchmarks_add += 1;
+                    let crate_name = times[0]["crate"].as_str().unwrap();
+                    data_benchmarks[index].by_crate.insert(test_name.to_string(),
+                        make_times(times, false).remove(crate_name).unwrap());
+                } else {
+                    c_benchmarks += 1;
+                    data_benchmarks.push(Data::new(date, &header, times, false));
+                }
+            }
+
+            for timing in times.members() {
+                for (phase, _) in timing["times"].entries() {
+                    phase_list.insert(phase.to_string());
+                }
+            }
+
+            if last_date.is_none() || last_date.as_ref().unwrap() < &date {
+                last_date = Some(date);
+            }
+        }
+
+        let last_date = last_date.expect("No dates found");
+
+        data_rustc.sort();
+        data_benchmarks.sort();
+
+        // Post processing to generate the summary data.
+        let summary_rustc = Summary::new(&data_rustc, last_date);
+        let summary_benchmarks = Summary::new(&data_benchmarks, last_date);
+
+        println!("read {} files", file_count);
+        println!("found {} files without times", no_times);
+        println!("{} bootstrap times", c_bootstraps);
+        println!("{} benchmarks times", c_benchmarks);
+        println!("{} benchmarks times (appended)", c_benchmarks_add);
+
+        Ok(ServerData {
+            summary_rustc: summary_rustc,
+            summary_benchmarks: summary_benchmarks,
+            crate_list: crate_list,
+            phase_list: phase_list,
+            benchmarks: benchmarks,
+            last_date: last_date,
+            data_rustc: data_rustc,
+            data_benchmarks: data_benchmarks,
+        })
+    }
+
+    pub fn by_kind(&self, kind: &str) -> &[Data] {
+        if kind == "rustc" {
+            &self.data_rustc
+        } else if kind == "benchmarks" {
+            &self.data_benchmarks
+        } else {
+            panic!("Unknown kind: {}", kind);
+        }
+    }
+}
+
+/// The data loaded for a single date, and all associated crates.
+pub struct Data {
```

Comment:
`TestRun` or `PerRun` or `PerRunData`, any of those or simialar seems fine to me (and better than Data).